python/python/lance/torch/data.py

```diff
@@ -111,7 +112,33 @@ def _to_tensor(
     return ret
 
 
-class TensorDataset(torch.utils.data.Dataset):
+try:
+    # When available, subclass from the newer torchdata DataPipes
+    # instead of torch Datasets.
+    import torchdata
+
+    MAP_DATASET_CLASS = torchdata.datapipes.map.MapDataPipe
+    ITER_DATASET_CLASS = torchdata.datapipes.iter.IterDataPipe
+
+    logger.warning(
+        "TorchData integration is still in BETA phase. "
+        "APIs may change without backward compatibility."
+    )
```

Comment:
```suggestion

    if "LANCE_TORCH_DATAPIPES" not in os.environ:
        logger.warning(
            "The torchdata version is old enough to utilize the DataPipe "
            "API but LANCE_TORCH_DATAPIPES is not set. We recommend "
            "either upgrading torchdata or setting the LANCE_TORCH_DATAPIPES "
            "environment variable in order to utilize torchdata"
        )
        raise ImportError("avoid deprecated data pipes API unless requested")
        
    logger.warning(
        "TorchData integration is still in BETA phase. "
        "APIs may change without backward compatibility."
    )
```

What if we guard the usage behind an environment variable.

If the user is using a newer version of `torchdata` they will not see the warning because the above `AttributeError` will be raised.
If the user does not have `torchdata` installed they will not see the warning because the `import torchdata` line will fail.
If the user does have an old version of `torchdata` we will first recommend they upgrade.
If the user is determined to use the old version of `torchdata` they can set the environment variable.